src/passes/RemoveNonJSOps.cpp

```diff
@@ -435,45 +301,1038 @@ struct RemoveNonJSOpsPass : public WalkerPass<PostWalker<RemoveNonJSOpsPass>> {
     Name functionCall;
     switch (curr->op) {
       case NearestFloat32:
-        needNearestF32 = true;
         functionCall = WASM_NEAREST_F32;
         break;
       case NearestFloat64:
-        needNearestF64 = true;
         functionCall = WASM_NEAREST_F64;
         break;
 
       case TruncFloat32:
-        needTruncF32 = true;
         functionCall = WASM_TRUNC_F32;
         break;
       case TruncFloat64:
-        needTruncF64 = true;
         functionCall = WASM_TRUNC_F64;
         break;
 
       case PopcntInt32:
-        needPopcntInt32 = true;
         functionCall = WASM_POPCNT32;
         break;
 
       case CtzInt32:
-        needCtzInt32 = true;
         functionCall = WASM_CTZ32;
         break;
 
       default: return;
     }
+    neededIntrinsics.insert(functionCall);
     replaceCurrent(builder->makeCall(functionCall, {curr->value}, curr->type));
   }
-
-private:
-  std::unique_ptr<Builder> builder;
 };
 
 Pass *createRemoveNonJSOpsPass() {
   return new RemoveNonJSOpsPass();
 }
 
+namespace {
+// A large WAST blob which contains the implementations of all the intrinsics
+// that we inject as part of this module. This blob was generated from a Rust
+// program [1] which uses the Rust compiler-builtins project. It's not
+// necessarily perfect but gets the job done! The idea here is that we inject
+// these pretty early so they can continue to be optimized by further passes
+// (aka inlining and whatnot)
+//
+// [1]: https://gist.github.com/alexcrichton/e7ea67bcdd17ce4b6254e66f77165690
+const char *IntrinsicsModule = R"(
```

Comment:
there was some discussion of options to put this in another file - did none of those pan out?

Reply:
Oops sorry, forgot to pull in those commits